File: aat-android/src/main/java/ch/bailu/aat/services/cache/ObjBitmap.java
package ch.bailu.aat.services.cache;

import org.mapsforge.core.graphics.Bitmap;

import ch.bailu.aat.services.ServiceContext;
import ch.bailu.aat_lib.service.background.FileTask;
import ch.bailu.aat.util.OldAppBroadcaster;
import ch.bailu.aat.util.graphic.SyncBitmap;
import ch.bailu.aat_lib.dispatcher.AppBroadcaster;
import ch.bailu.aat_lib.service.cache.Obj;
import ch.bailu.foc.Foc;
import ch.bailu.foc_android.FocAndroid;

public final class ObjBitmap extends ObjImageAbstract {
    public final static ObjBitmap NULL=new ObjBitmap();

    private final SyncBitmap bitmap=new SyncBitmap();
    private final Foc imageFile;


    private ObjBitmap() {
        this(FocAndroid.NULL);
    }

    public ObjBitmap(Foc id) {
        super(id.getPath());
        imageFile = id;
    }




    @Override
    public void onInsert(ServiceContext sc){
        load(sc);

        sc.getCacheService().addToBroadcaster(this);
    }


    public void onRemove(ServiceContext sc) {
        super.onRemove(sc);
        bitmap.free();
    }



    private void load(ServiceContext sc) {
        sc.getBackgroundService().process(new BitmapLoader(imageFile));
    }



    @Override
    public long getSize() {
        return bitmap.getSize();
    }

    @Override
    public boolean isReadyAndLoaded() {
        return getBitmap() != null;
    }

    @Override
    public synchronized Bitmap getBitmap() {
        return bitmap.getBitmap();
    }


    public static class Factory extends Obj.Factory {
        @Override
        public Obj factory(String id, ServiceContext sc) {
            return new ObjBitmap(FocAndroid.factory(sc.getContext(),id));
        }
    }


    @Override
    public void onDownloaded(String id, String url, ServiceContext sc) {
        if (id.equals(toString())) {
            load(sc);
        }
    }


    @Override
    public void onChanged(String id, ServiceContext sc) {}



    private static class BitmapLoader extends FileTask {

        public BitmapLoader(Foc f) {
            super(f);
        }

        @Override
        public long bgOnProcess(final ServiceContext sc) {
            final long[] size = {0};

            new OnObject(sc, toString(), ObjBitmap.class) {
                @Override
                public void run(Obj obj) {
                    ObjBitmap self = (ObjBitmap) obj;

                    try {
                        self.bitmap.set(self.imageFile);
                        size[0] =  self.bitmap.getSize();
                    } catch (Exception e) {
                        self.setException(e);
                    }

                    OldAppBroadcaster.broadcast(sc.getContext(), AppBroadcaster.FILE_CHANGED_INCACHE,
                            self.imageFile);

                }
            };
            return size[0];
        }

    }


}

File: aat-android/src/main/kotlin/ch/bailu/aat/services/cache/ObjBitmap.kt
package ch.bailu.aat.services.cache

import ch.bailu.aat.util.graphic.SyncBitmap
import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.dispatcher.AppBroadcaster
import ch.bailu.aat_lib.service.ServicesInterface
import ch.bailu.aat_lib.service.background.FileTask
import ch.bailu.aat_lib.service.cache.Obj
import ch.bailu.aat_lib.service.cache.OnObject
import ch.bailu.aat_lib.service.cache.icons.ObjImageAbstract
import ch.bailu.foc.Foc
import org.mapsforge.core.graphics.Bitmap

class ObjBitmap(private val imageFile: Foc) : ObjImageAbstract(imageFile.path) {
    private val syncBitmap = SyncBitmap()

    private constructor() : this(Foc.FOC_NULL)

    override fun onInsert(sc: AppContext) {
        load(sc.services)
        sc.services.cacheService.addToBroadcaster(this)
    }

    override fun onRemove(sc: AppContext) {
        super.onRemove(sc)
        syncBitmap.free()
    }

    private fun load(sc: ServicesInterface) {
        sc.backgroundService.process(BitmapLoader(imageFile))
    }

    override fun getSize(): Long {
        return syncBitmap.size.toLong()
    }

    override fun isReadyAndLoaded(): Boolean {
        return syncBitmap.bitmap != null
    }

    override fun getBitmap(): Bitmap? {
        return syncBitmap.bitmap
    }

    class Factory : Obj.Factory() {
        override fun factory(id: String, sc: AppContext): Obj {
            return ObjBitmap(sc.toFoc(id))
        }
    }

    override fun onDownloaded(id: String, url: String, sc: AppContext) {
        if (id == toString()) {
            load(sc.services)
        }
    }

    override fun onChanged(id: String, sc: AppContext) {}
    private class BitmapLoader(f: Foc) : FileTask(f) {

        override fun bgOnProcess(appContext: AppContext): Long {

            var size = 0L
            object : OnObject(appContext, toString(), ObjBitmap::class.java) {
                override fun run(obj: Obj) {
                    val self = obj as ObjBitmap
                    try {
                        self.syncBitmap.set(self.imageFile)
                        size = self.syncBitmap.size.toLong()
                    } catch (e: Exception) {
                        self.exception = e
                    }
                    appContext.broadcaster.broadcast(
                        AppBroadcaster.FILE_CHANGED_INCACHE,
                        self.imageFile.path
                    )
                }
            }
            return size
        }
    }

    companion object {
        val NULL = ObjBitmap()
    }
}
